src/Umbraco.Core/DependencyInjection/UmbracoBuilder.Events.cs

```diff
@@ -26,5 +27,22 @@ public static partial class UmbracoBuilderExtensions
             builder.Services.AddTransient(typeof(INotificationHandler<TNotification>), typeof(TNotificationHandler));
             return builder;
         }
+
+        /// <summary>
+        /// Registers a notification handler against the Umbraco service collection.
+        /// </summary>
+        /// <typeparam name="TNotification">The type of notification.</typeparam>
+        /// <typeparam name="TNotificationHandler">The type of notificiation handler.</typeparam>
+        /// <param name="builder">The Umbraco builder.</param>
+        /// <param name="factory">Factory method</param>
+        /// <returns>The <see cref="IUmbracoBuilder"/>.</returns>
+        public static IUmbracoBuilder AddNotificationHandler<TNotification, TNotificationHandler>(this IUmbracoBuilder builder, Func<IServiceProvider, TNotificationHandler> factory)
+            where TNotificationHandler : class, INotificationHandler<TNotification>
+            where TNotification : INotification
+        {
+            // Register the handler as transient. This ensures that anything can be injected into it.
+            builder.Services.AddTransient(typeof(INotificationHandler<TNotification>), factory);
```

Comment:
Wondering if this should be TryAddTransient so that duplicated notifications/handler combinations are not accidentally added? I don't think it would be expected that the same notification/handler combination should exist more than once?

Reply:
Nope. Because that will prevent more than one handler being registered at all since they’re registered against the interface.

Reply:
Try it and watch the unit tests fail

Reply:
Roger that

Reply:
Presumably you can achieve the same with replace and an explicit service descriptor, but I’m not sure I get the point. Is it just trying to save someone from shooting them selves in the foot?

who’s going to add same handler twice?

Reply:
I don't think anyone on purpose will add the same handler twice but since handlers are going to be registered with our "Add..." ext methods, there's nothing preventing people from calling these ext methods twice. 

I've seen places in libraries that will call `Add` methods that the library's functionality needs without a care of whether it's been called before or not (i.e. `AddHttpClient`) and I think this will inevitably be done with some of our ext methods too.  

In the case of event handlers, it would not be good to accidentally have multiple event handlers of the same handler executing.

> Is it just trying to save someone from shooting them selves in the foot?

Indeed, people are quite good at shooting their own feet :P 

Maybe something to consider later as the code progresses with the migration of core composers to ext methods. Maybe in the end it might not be a worry, lets see how it goes.

Reply:
This could be solved by checking this and tests still pass

```cs
var descriptor = new ServiceDescriptor(typeof(INotificationHandler<TNotification>), typeof(TNotificationHandler), ServiceLifetime.Transient);
if (!builder.Services.Contains(descriptor))
{   
    builder.Services.Add(descriptor);
}
```

An example of why this might be needed:

* in https://github.com/umbraco/Umbraco-CMS/pull/9598 there's a new ext method `AddDistributedCache` to replace a composer for distributed cache stuff
* both the back office and front-end website require this to be called
* but we want the ability to have either the back office or front-end registered independently, or both
* currently that is done with `.AddBackOffice()` and `.AddWebsite()`
* in theory, both of these methods would also call `AddDistributedCache` since they both need it, alternatively this would need to be called as an additional method which is a bit annoying

Of course there's other ways to resolve this too based on what ext methods we want people to use to add the required services while trying to make it simple for end users. 

For now I've used the above code but happy to change based on feedback.

Reply:
Does `AddDistributedCache` have to be an extension on `IUmbracoBuilder`? Remember that we want to expose the absolute minimum on that interface to general consumers.  

That fix looks fine to me btw. 👍

Reply:
> Does AddDistributedCache have to be an extension on IUmbracoBuilder? Remember that we want to expose the absolute minimum on that interface to general consumers.

That's just currently how it's done but as we progress with the front-end rendering stuff a lot will be updated. I agree and shouldn't expose things publicly that aren't necessary. It can already be internal but since both the back office and front-end rendering require this it's more an 'ensure' services type of call. There will probably be others like this too. Totally open for suggestions but I guess there could be a few solutions. One would be to have a different root ext method for running umbraco as a website (as opposed to booting for things like integration tests, or non-web apps)

```cs

// add all of umbraco, back office and front-end
.AddUmbracoWebApplication() // core + shared things like dist cache and other 
  .AddBackOffice()
  .AddWebsite();

// add only back office
.AddUmbracoWebApplication()
  .AddBackOffice();

// add only front end
.AddUmbracoWebApplication()
  .AddWebsite();
```

Then perhaps UseUmbracoBackOffice/UseUmbracoWebsite do some service validation checks [like UseMvc does](https://github.com/dotnet/aspnetcore/blob/b795ac3546eb3e2f47a01a64feb3020794ca33bb/src/Mvc/Mvc.Core/src/Builder/MvcApplicationBuilderExtensions.cs#L132)